test/lit/passes/dae-refine-params-and-optimize.wast

```diff
@@ -0,0 +1,31 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+
+;; RUN: foreach %s %t wasm-opt -all --dae-optimizing -S -o - | filecheck %s
+
+(module
+ ;; CHECK:      (type $ref|array|_=>_i32 (func (param (ref array)) (result i32)))
+
+ ;; CHECK:      (func $len (type $ref|array|_=>_i32) (param $0 (ref array)) (result i32)
+ ;; CHECK-NEXT:  (drop
+ ;; CHECK-NEXT:   (return
+ ;; CHECK-NEXT:    (array.len
+ ;; CHECK-NEXT:     (local.get $0)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $len (param (ref eq)) (result i32)
+  (drop
```

Comment:
Maybe worth a comment here about what is expected and what is not. Maybe:

"The `br` is optimized away thanks to the refined type, but the output still has some unoptimized code (`--dce --remove-unused-brs` would get rid of the `drop-return` pattern here, but that is unrelated to this test)."

Reply:
I was mostly interested in that compiling with `-O3` does not produce never-branching `br_on_cast_fail` (since they are crashing V8).

But I'm wondering about this remaining `drop`.  What happens is that the `remove-unused-brs` pass is scheduled after the `dce` pass in `PassRunner::addDefaultFunctionOptimizationPasses`, even though it can expose some dead code.

Reply:
Yeah, this does show that the current ordering of passes is not optimal. But I'm not sure we can improve it without increasing compile times (by adding another run of `--dce`). This might be worth thinking about, but in general my feeling is that GC code will want multiple iterations of optimization anyhow, like `-O3 -O3`, which would handle this in most cases.  But at some point perhaps looking into a more efficient pass ordering could make sense.